ejb3/src/main/java/org/jboss/as/ejb3/tx/CMTTxInterceptor.java

```diff
@@ -252,41 +175,68 @@ protected Object invokeInCallerTx(InterceptorContext invocation, Transaction tx,
         try {
             return invocation.proceed();
         } catch (Throwable t) {
-            handleInCallerTx(invocation, t, tx, component);
+            ApplicationExceptionDetails ae = component.getApplicationException(t.getClass(), invocation.getMethod());
+
+            if (ae != null) {
+                if (ae.isRollback()) setRollbackOnly(tx);
+                // an app exception can never be an Error
+                throw (Exception) t;
+            }
+            try {
+                throw t;
+            } catch (EJBTransactionRolledbackException | NoSuchEJBException | NoSuchEntityException e) {
+                setRollbackOnly(tx);
+                throw e;
+            } catch (RuntimeException e) {
+                setRollbackOnly(tx);
+                throw new EJBTransactionRolledbackException(e.getMessage(), e);
+            } catch (Error e) {
+                setRollbackOnly(tx);
+                throw EjbLogger.ROOT_LOGGER.unexpectedErrorRolledBack(e);
+            }
         }
-        throw new RuntimeException("UNREACHABLE");
     }
 
     protected Object invokeInNoTx(InterceptorContext invocation, final EJBComponent component) throws Exception {
         try {
             return invocation.proceed();
-        } catch (Throwable t) {
-            handleExceptionInNoTx(invocation, t, component);
+        } catch (EJBException | RemoteException e) {
+            throw e;
+        } catch (RuntimeException e) {
+            throw new EJBException(e);
```

Comment:
Previously if a RuntimeException was listed as an application exception it would not be wrapped, which is the behaviour expected by the spec.

Reply:
OK so that transforms into:

```java
        try {
            return invocation.proceed();
        } catch (EJBException | RemoteException e) {
            throw e;
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw e;
        } catch (Error e) {
            ApplicationExceptionDetails ae = component.getApplicationException(e.getClass(), invocation.getMethod());
            if (ae != null) {
                throw e;
            } else {
                throw EjbLogger.ROOT_LOGGER.unexpectedError(e);
            }
        } catch (Throwable t2) {
            throw new UndeclaredThrowableException(t2);
        }
```

Then collapsing catch blocks gives us:

```java
        try {
            return invocation.proceed();
        } catch (Exception e) {
            throw e;
        } catch (Error e) {
            ApplicationExceptionDetails ae = component.getApplicationException(e.getClass(), invocation.getMethod());
            if (ae != null) {
                throw e;
            } else {
                throw EjbLogger.ROOT_LOGGER.unexpectedError(e);
            }
        } catch (Throwable t2) {
            throw new UndeclaredThrowableException(t2);
        }
```

Take into account that AED can't be an error and we get:

```java
        try {
            return invocation.proceed();
        } catch (Exception e) {
            throw e;
        } catch (Error e) {
            throw EjbLogger.ROOT_LOGGER.unexpectedError(e);
        } catch (Throwable t2) {
            throw new UndeclaredThrowableException(t2);
        }
```

That seems OK, right?  I wonder if I should delete the catch for ``Throwable`` there too.

Reply:
Making this change regresses the AppExceptionNever test.

The old code did this:

```java
        // if it's neither EJBException nor RemoteException
        if (!(t instanceof EJBException || t instanceof RemoteException)) {
            // errors and unchecked are wrapped into EJBException
            if (t instanceof Error) {
                //t = new EJBException(formatException("Unexpected Error", t));
                Throwable cause = t;
                t = EjbLogger.ROOT_LOGGER.unexpectedError();
                t.initCause(cause);
            } else if (t instanceof RuntimeException) {
                t = new EJBException((Exception) t);
            } else {
                // an application exception
                throw (Exception) t;
            }
        }

        throw (Exception) t;
```

So in this case RuntimeException would pass the ``if`` check and would get wrapped in EJBException.

Reply:
If the spec is forbidding this behavior then we need to fix the test.

Reply:
OK disregard that last stuff.  Re-analyzing the old code, it will only wrap RuntimeExceptions which are not EJBException or application exceptions.

Errors always get wrapped (because they can never be app exceptions), and checked Exceptions never get wrapped.  Updated fix is coming in a moment.